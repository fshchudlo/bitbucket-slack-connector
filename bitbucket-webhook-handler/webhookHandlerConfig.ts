import { BitbucketNotification } from "../typings";

export interface WebhookHandlerConfig {
    readonly usePrivateChannels: boolean;
    readonly defaultChannelParticipants?: string[];
    getOpenedPRBroadcastChannelId(payload: BitbucketNotification): string | null;
}