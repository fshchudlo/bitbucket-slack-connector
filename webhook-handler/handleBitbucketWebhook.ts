import {
    sendCompletionMessageAndCloseTheChannel,
    setChannelTopicAndInviteParticipants,
    sendMessageAboutAddedComment,
    updateChannelMembers,
    sendMessageAboutNewCommit,
    sendMessageAboutReviewerAction,
    sendMessageAboutDeletedComment, sendMessageAboutEditedComment
} from "./use-cases";
import { SlackAPIAdapter } from "./SlackAPIAdapter";
import { BitbucketGateway } from "./BitbucketGateway";
import { sendMessageAboutPRModification } from "./use-cases/sendMessageAboutPRModification";
import { BitbucketNotification } from "../typings";
import { buildChannelName } from "./slack-helpers";

export default async function handleBitbucketWebhook(payload: BitbucketNotification, slackAPI: SlackAPIAdapter, bitbucketGateway: BitbucketGateway, usePrivateChannels: boolean = true) {
    const eventKey = payload.eventKey;
    const channelInfo = await provisionPullRequestChannel(slackAPI, payload, usePrivateChannels);
    switch (eventKey) {
        case "pr:opened":
            await setChannelTopicAndInviteParticipants(payload, slackAPI, channelInfo);
            break;
        case "pr:modified":
            await sendMessageAboutPRModification(payload, slackAPI, channelInfo);
            break;
        case "pr:reviewer:updated":
            await updateChannelMembers(payload, slackAPI, channelInfo);
            break;
        case "pr:reviewer:unapproved":
        case "pr:reviewer:needs_work":
        case "pr:reviewer:approved":
            await sendMessageAboutReviewerAction(payload, slackAPI, channelInfo);
            break;
        case "pr:comment:added":
            await sendMessageAboutAddedComment(payload, slackAPI, channelInfo);
            break;
        case "pr:comment:edited":
            await sendMessageAboutEditedComment(payload, slackAPI, channelInfo);
            break;
        case "pr:comment:deleted":
            await sendMessageAboutDeletedComment(payload, slackAPI, channelInfo);
            break;
        case "pr:from_ref_updated":
            await sendMessageAboutNewCommit(payload, slackAPI, bitbucketGateway, channelInfo);
            break;
        case "pr:merged":
        case "pr:declined":
        case "pr:deleted":
            await sendCompletionMessageAndCloseTheChannel(payload, slackAPI, channelInfo);
            break;
        default:
            throw new Error(`"${eventKey}" event key is unknown.`);
    }
}

async function provisionPullRequestChannel(slackAPI: SlackAPIAdapter, payload: BitbucketNotification, usePrivateChannels: boolean) {
    return payload.eventKey == "pr:opened" ? await slackAPI.createChannel({
        name: buildChannelName(payload.pullRequest),
        isPrivate: usePrivateChannels
    }) : await slackAPI.findChannel(buildChannelName(payload.pullRequest), true);
}
