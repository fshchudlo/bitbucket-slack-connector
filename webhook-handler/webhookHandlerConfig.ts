export type WebhookHandlerConfig = {
    readonly USE_PRIVATE_CHANNELS: boolean;
    readonly DEFAULT_CHANNEL_PARTICIPANTS?: string[];
    readonly BROADCAST_OPENED_PR_MESSAGES_TO_CHANNEL_ID?: string;
}