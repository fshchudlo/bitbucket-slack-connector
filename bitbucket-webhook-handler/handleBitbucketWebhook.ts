import * as useCases from "./use-cases";
import { SlackAPIAdapter } from "./ports/SlackAPIAdapter";
import { BitbucketAPIAdapter } from "./ports/BitbucketAPIAdapter";
import { BitbucketNotification } from "../bitbucket-payload-types";
import { WebhookHandlerConfig } from "./webhookHandlerConfig";
import { provisionPullRequestChannel } from "./provisionPullRequestChannel";

export default async function handleBitbucketWebhook(payload: BitbucketNotification, slackAPI: SlackAPIAdapter, bitbucketAPI: BitbucketAPIAdapter, config: WebhookHandlerConfig) {
    const eventKey = payload.eventKey;
    const channelInfo = await provisionPullRequestChannel(slackAPI, bitbucketAPI, payload, config);

    switch (eventKey) {
        case "pr:opened":
            await useCases.inviteParticipantsAndSetChannelBookmark(payload, slackAPI, config.defaultChannelParticipants, channelInfo.id);
            await useCases.tryBroadcastMessageAboutOpenedPR(payload, slackAPI, config.getOpenedPRBroadcastChannelId(payload));
            break;
        case "pr:modified":
            await useCases.sendMessageAboutPRModification(payload, slackAPI, channelInfo.id);
            break;
        case "pr:reviewer:updated":
            await useCases.updateChannelMembers(payload, slackAPI, channelInfo.id);
            break;
        case "pr:reviewer:unapproved":
        case "pr:reviewer:needs_work":
        case "pr:reviewer:approved":
            await useCases.sendMessageAboutReviewerAction(payload, slackAPI, channelInfo.id);
            break;
        case "pr:comment:added":
            await useCases.sendMessageAboutAddedComment(payload, slackAPI, channelInfo.id);
            break;
        case "pr:comment:edited":
            await useCases.sendMessageAboutEditedComment(payload, slackAPI, channelInfo.id);
            break;
        case "pr:comment:deleted":
            await useCases.sendMessageAboutDeletedComment(payload, slackAPI, channelInfo.id);
            break;
        case "pr:from_ref_updated":
            await useCases.sendMessageAboutNewCommit(payload, slackAPI, bitbucketAPI, channelInfo.id);
            break;
        case "pr:merged":
        case "pr:declined":
        case "pr:deleted":
            await useCases.sendCompletionMessageAndCloseTheChannel(payload, slackAPI, channelInfo.id);
            await useCases.tryBroadcastMessageAboutClosedPR(payload, slackAPI, config.getOpenedPRBroadcastChannelId(payload));
            break;
        default:
            throw new Error(`"${eventKey}" event key is unknown.`);
    }
}

