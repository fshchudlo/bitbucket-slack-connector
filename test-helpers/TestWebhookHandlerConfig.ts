import { WebhookHandlerConfig } from "../bitbucket-webhook-handler/webhookHandlerConfig";

export const TestWebhookHandlerConfig: WebhookHandlerConfig = {
    getOpenedPRBroadcastChannelId(): string | null {
        return null;
    },
    usePrivateChannels: true
};