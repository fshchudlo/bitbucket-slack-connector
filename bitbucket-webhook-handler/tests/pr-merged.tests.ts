import SlackAdapterSnapshottingMock from "./mocks/SlackAdapterSnapshottingMock";
import TestPayloadBuilder from "./mocks/TestPayloadBuilder";
import handleBitbucketWebhook from "../handleBitbucketWebhook";
import { MockBitbucketAPIAdapter } from "./mocks/MockBitbucketAPIAdapter";
import { TestWebhookHandlerConfig } from "./mocks/TestWebhookHandlerConfig";
import { WebhookHandlerConfig } from "../webhookHandlerConfig";

describe("handleBitbucketWebhook", () => {
    it("Should send completion message and close the channel on PR merge", async () => {
        const testSlackGateway = await new SlackAdapterSnapshottingMock().setupBasicChannel();

        await handleBitbucketWebhook(TestPayloadBuilder.pullRequestMerged(), testSlackGateway, new MockBitbucketAPIAdapter(), TestWebhookHandlerConfig);

        expect(testSlackGateway.snapshot).toMatchSnapshot();
    });

    it("Should send notification to the broadcast channel, if it is specified", async () => {
        const testConfig:WebhookHandlerConfig = {
            ...TestWebhookHandlerConfig,
            getOpenedPRBroadcastChannelId: () => "test-broadcast-channel-id"
        };
        const testSlackGateway = await new SlackAdapterSnapshottingMock().setupBasicChannel(testConfig);


        await handleBitbucketWebhook(TestPayloadBuilder.pullRequestMerged(), testSlackGateway, new MockBitbucketAPIAdapter(), testConfig);


        expect(testSlackGateway.snapshot).toMatchSnapshot();
    });

});
