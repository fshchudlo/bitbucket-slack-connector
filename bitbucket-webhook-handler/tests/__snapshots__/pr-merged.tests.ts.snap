// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleBitbucketWebhook Should send completion message and close the channel on PR merge 1`] = `
{
  "addedBookmarks": [],
  "addedReactions": [],
  "archivedChannels": [
    "12345",
  ],
  "createdChannels": [
    {
      "id": "12345",
      "isArchived": false,
      "name": "pr-test-proj-test-repository-1",
    },
  ],
  "invitesToChannels": [],
  "kicksFromChannels": [],
  "lookedUpUsers": [],
  "searchedChannels": [],
  "searchedCommentSnapshots": [],
  "searchedPrOpenedBroadcastMessages": [],
  "sentMessages": [
    {
      "channelId": "12345",
      "iconEmoji": ":bitbucket:",
      "text": ":white_check_mark: Pull request was merged by Test Author. Well done, thank you all.",
    },
  ],
}
`;

exports[`handleBitbucketWebhook Should send notification to the broadcast channel, if it is specified 1`] = `
{
  "addedBookmarks": [],
  "addedReactions": [],
  "archivedChannels": [
    "12345",
  ],
  "createdChannels": [
    {
      "id": "12345",
      "isArchived": false,
      "name": "pr-test-proj-test-repository-1",
    },
  ],
  "invitesToChannels": [],
  "kicksFromChannels": [],
  "lookedUpUsers": [],
  "searchedChannels": [],
  "searchedCommentSnapshots": [],
  "searchedPrOpenedBroadcastMessages": [
    {
      "channelId": "test-broadcast-channel-id",
      "prCreationDate": 2024-04-29T08:59:44.802Z,
      "pullRequestTraits": {
        "projectKey": "TEST-PROJ",
        "pullRequestId": "1",
        "repositorySlug": "test-repository",
      },
    },
  ],
  "sentMessages": [
    {
      "channelId": "12345",
      "iconEmoji": ":bitbucket:",
      "text": ":white_check_mark: Pull request was merged by Test Author. Well done, thank you all.",
    },
  ],
}
`;
